File: theses.py
"""Investment-thesis framework — shared helpers.

When an agent buys an equity, ``PortfolioManager.buy()`` /
``buy_atomic()`` call ``record_thesis()`` here so the system captures a
durable record of the trade. Two tiers:

* **Snapshot** (always populated) — a frozen JSONB capture of the
  equity's state at buy time. Shape mirrors the ``extended`` tier of
  ``build_universe_snapshot.py`` (fundamentals + valuation + momentum
  + narrative). See ``build_snapshot``.

* **Thesis text + signals** (optional, agent-supplied) — narrative and
  machine-checkable extend/break signals. NULL when the buy call
  passed no ``thesis`` kwarg.

Lifecycle: ``active`` (default) → ``broken`` / ``improved`` /
``superseded`` / ``closed`` (terminal). ``superseded`` flows
automatically when the same agent buys the same ticker again;
``closed`` flows when ``close_theses_for_position`` is called after
a full position exit. ``broken`` / ``improved`` are caller-managed —
``check_thesis`` is read-only and returns a verdict; agents that want
to persist that verdict call ``mark_thesis_status``.

Signal operators supported by ``check_thesis``:

    >, >=, <, <=, ==, !=        — current vs static value
    change_pct_lt, change_pct_gt — current vs snapshot, percentage delta

The change_pct variants compare current vs the value captured in the
snapshot. ``{"op": "change_pct_lt", "value": -5}`` means *"current is
more than 5 percentage points below the snapshot value"* (i.e. a
collapse). Use the regular comparison operators for static thresholds.
"""

from __future__ import annotations

import logging
from typing import Any, Optional

logger = logging.getLogger("theses")


# Fields captured into the ``snapshot`` JSONB at buy time. Mirrors the
# ``extended`` tier of build_universe_snapshot.py (fundamentals +
# valuation + momentum + narrative + a few audit fields).
_SNAPSHOT_FIELDS: tuple[str, ...] = (
    # Identity / overview
    "ticker", "company_name", "country", "sector",
    # Fundamentals (extended tier)
    "rating", "r40_score", "rule_of_40",
    "rev_growth_ttm_pct", "rev_growth_qoq_pct", "rev_cagr_pct",
    "rev_consistency_score",
    "gross_margin_pct", "operating_margin_pct", "net_margin_pct",
    "net_margin_yoy_pct", "fcf_margin_pct",
    "opex_pct_revenue", "sm_rd_pct_revenue",
    "eps_only", "eps_yoy_pct", "qrtrs_to_profitability",
    "gm_trend",
    # Valuation
    "price", "ps_now", "price_pct_of_52w_high",
    # Momentum
    "perf_52w_vs_spy", "composite_score",
    # Narrative
    "short_outlook", "key_risks", "full_outlook", "bull_eval", "bear_eval",
    "status",
    # Quality flags + audit
    "flags", "ai_analyzed_at",
)


def build_snapshot(db, ticker: str) -> dict:
    """Read the latest equity state and return the dict for the
    ``snapshot`` JSONB column.

    Returns a dict containing exactly the fields in ``_SNAPSHOT_FIELDS``
    (missing fields become None). Raises ``ValueError`` if the ticker
    isn't in ``companies``.
    """
    company = db.get_company(ticker)
    if not company:
        raise ValueError(f"no companies row for {ticker}")
    return {k: company.get(k) for k in _SNAPSHOT_FIELDS}


def record_thesis(
    db,
    *,
    agent_id: str,
    ticker: str,
    trade_id: Optional[int] = None,
    thesis_text: Optional[str] = None,
    extend_signals: Optional[list[dict]] = None,
    break_signals: Optional[list[dict]] = None,
) -> int:
    """Insert an investment_theses row for a freshly-executed BUY.

    Always captures the snapshot. Marks any prior ``active`` row for
    the same (agent_id, ticker) as ``superseded``. Returns the new
    thesis id.

    ``source='agent'`` when any of thesis_text / extend_signals /
    break_signals is provided; otherwise ``source='auto'``.
    """
    snapshot = build_snapshot(db, ticker)

    # Mark any prior active thesis for this (agent, ticker) as superseded.
    db.client.table("investment_theses").update(
        {"status": "superseded", "status_changed_at": "now()"}
    ).match(
        {"agent_id": agent_id, "ticker": ticker, "status": "active"}
    ).execute()

    source = "agent" if (
        thesis_text or extend_signals or break_signals
    ) else "auto"

    row = {
        "agent_id": agent_id,
        "ticker": ticker,
        "trade_id": trade_id,
        "snapshot": snapshot,
        "thesis_text": thesis_text,
        "extend_signals": extend_signals,
        "break_signals": break_signals,
        "source": source,
        "status": "active",
    }
    resp = db.client.table("investment_theses").insert(row).execute()
    inserted = (resp.data or [{}])[0]
    new_id = inserted.get("id")
    logger.info(
        "thesis %s recorded: agent=%s ticker=%s source=%s trade_id=%s",
        new_id, agent_id[:8] if agent_id else "?", ticker, source, trade_id,
    )
    return new_id


def close_theses_for_position(db, *, agent_id: str, ticker: str) -> int:
    """Flip all non-closed theses for (agent_id, ticker) to ``closed``.

    Called from ``PortfolioManager.sell`` after a sell zeros out the
    holding. Idempotent — if there are no matching rows (e.g. older
    positions opened before this migration landed), it's a no-op.

    Returns the number of rows updated.
    """
    resp = (
        db.client.table("investment_theses")
        .update({
            "status": "closed",
            "status_changed_at": "now()",
            "closed_at": "now()",
        })
        .match({"agent_id": agent_id, "ticker": ticker})
        .neq("status", "closed")
        .execute()
    )
    rows = resp.data or []
    if rows:
        logger.info(
            "closed %d theses for agent=%s ticker=%s",
            len(rows), agent_id[:8] if agent_id else "?", ticker,
        )
    return len(rows)


# ---------------------------------------------------------------------------
# Maintenance check — pure read, no side effects
# ---------------------------------------------------------------------------


def _coerce_number(v: Any) -> Optional[float]:
    """Best-effort float conversion. None / em-dash / unparseable → None."""
    if v is None or v == "—":
        return None
    try:
        return float(v)
    except (TypeError, ValueError):
        return None


_STATIC_OPS = {
    ">":  lambda c, t: c >  t,
    ">=": lambda c, t: c >= t,
    "<":  lambda c, t: c <  t,
    "<=": lambda c, t: c <= t,
    "==": lambda c, t: c == t,
    "!=": lambda c, t: c != t,
}


def _evaluate_signal(signal: dict, snapshot: dict, current: dict) -> bool:
    """Return True if the signal is triggered, False otherwise.

    A signal is a dict ``{field, op, value, description?}``. Unknown
    operators, missing fields, or non-numeric values all return False
    (a malformed or unevaluable signal is treated as not-yet-triggered
    — conservative, agents see no false positives).
    """
    field = signal.get("field")
    op = signal.get("op")
    threshold = signal.get("value")
    if not field or not op:
        return False
    current_value = _coerce_number(current.get(field))
    if current_value is None:
        return False
    if op in _STATIC_OPS:
        threshold_num = _coerce_number(threshold)
        if threshold_num is None:
            return False
        return _STATIC_OPS[op](current_value, threshold_num)
    if op in ("change_pct_lt", "change_pct_gt"):
        snapshot_value = _coerce_number(snapshot.get(field))
        threshold_num = _coerce_number(threshold)
        if snapshot_value is None or threshold_num is None:
            return False
        # Percentage-point delta (current - snapshot), not relative %
        delta_pp = current_value - snapshot_value
        if op == "change_pct_lt":
            return delta_pp < threshold_num
        return delta_pp > threshold_num
    return False


def check_thesis(db, thesis_id: int) -> dict:
    """Compare a thesis's snapshot + signals against current state.

    Pure read — does NOT mutate the row. Caller decides what to do
    with the verdict (e.g. call ``mark_thesis_status`` to persist).

    Returns::

        {
          "verdict": "active" | "broken" | "improved",
          "broken_signals":           [<signal>, ...],
          "confirmed_extend_signals": [<signal>, ...],
          "delta": {<field>: {"snapshot": v, "current": v}, ...},
        }

    Precedence: any triggered break_signal → ``"broken"``; else any
    triggered extend_signal → ``"improved"``; else ``"active"``.
    For ``source='auto'`` rows with no signals the verdict is always
    ``"active"`` — the snapshot delta is still computed so callers can
    inspect drift.
    """
    resp = (
        db.client.table("investment_theses")
        .select("*")
        .eq("id", thesis_id)
        .execute()
    )
    rows = resp.data or []
    if not rows:
        raise ValueError(f"no investment_theses row with id={thesis_id}")
    thesis = rows[0]

    ticker = thesis["ticker"]
    snapshot = thesis.get("snapshot") or {}
    current_row = db.get_company(ticker) or {}

    break_signals = thesis.get("break_signals") or []
    extend_signals = thesis.get("extend_signals") or []

    triggered_breaks = [
        s for s in break_signals if _evaluate_signal(s, snapshot, current_row)
    ]
    triggered_extends = [
        s for s in extend_signals if _evaluate_signal(s, snapshot, current_row)
    ]

    if triggered_breaks:
        verdict = "broken"
    elif triggered_extends:
        verdict = "improved"
    else:
        verdict = "active"

    # Snapshot delta — useful even when there are no signals.
    delta = {}
    for field in _SNAPSHOT_FIELDS:
        s_val = snapshot.get(field)
        c_val = current_row.get(field)
        if s_val != c_val:
            delta[field] = {"snapshot": s_val, "current": c_val}

    return {
        "verdict": verdict,
        "broken_signals": triggered_breaks,
        "confirmed_extend_signals": triggered_extends,
        "delta": delta,
    }


def mark_thesis_status(
    db, thesis_id: int, *, status: str, reason: Optional[str] = None,
) -> None:
    """Persist a status change. Optional — callers can also rely on
    ``check_thesis`` as a read-only oracle without ever writing back.

    ``status`` must be one of ``active``, ``broken``, ``improved``,
    ``superseded``, ``closed``. ``reason`` is logged but not stored
    (no column for it yet — add later if needed).
    """
    valid = {"active", "broken", "improved", "superseded", "closed"}
    if status not in valid:
        raise ValueError(f"status must be one of {valid}, got {status!r}")
    fields = {"status": status, "status_changed_at": "now()"}
    if status == "closed":
        fields["closed_at"] = "now()"
    db.client.table("investment_theses").update(fields).match(
        {"id": thesis_id}
    ).execute()
    if reason:
        logger.info("thesis %s → %s (%s)", thesis_id, status, reason)
    else:
        logger.info("thesis %s → %s", thesis_id, status)

File: web/lib/theses.ts
/**
 * TypeScript port of the investment-thesis recording path.
 *
 * Mirrors `theses.py` (the Python helper module used by house agents
 * running through `agent_heartbeat.py`). Together with the changes in
 * `web/lib/portfolio.ts`, this makes the snapshot capture **universal**:
 * every successful BUY records an `investment_theses` row regardless
 * of which code path executed the trade.
 *
 * The full Python module also exposes `check_thesis` / `mark_thesis_status`
 * for the maintenance check loop — those aren't needed by the public
 * buy/sell endpoints so they're intentionally not ported here. Read-side
 * access to investment_theses on the website goes through
 * `web/lib/theses-query.ts`.
 *
 * See `migrations/020_investment_theses.sql` for the table contract.
 */

import { getSupabase } from "@/lib/supabase";
import { getEquityL0 } from "@/lib/level0-query";

export interface ThesisSignal {
  field: string;
  op: string;
  value: number | string;
  description?: string;
}

export interface ThesisInput {
  thesis_text?: string | null;
  extend_signals?: ThesisSignal[] | null;
  break_signals?: ThesisSignal[] | null;
}

// Fields captured into the `snapshot` JSONB at buy time. Must stay in
// lock-step with `_SNAPSHOT_FIELDS` in theses.py — keep them sorted the
// same way so a diff is easy to read.
const SNAPSHOT_FIELDS = [
  // Identity / overview
  "ticker", "company_name", "country", "sector",
  // Fundamentals (extended tier)
  "rating", "r40_score", "rule_of_40",
  "rev_growth_ttm_pct", "rev_growth_qoq_pct", "rev_cagr_pct",
  "rev_consistency_score",
  "gross_margin_pct", "operating_margin_pct", "net_margin_pct",
  "net_margin_yoy_pct", "fcf_margin_pct",
  "opex_pct_revenue", "sm_rd_pct_revenue",
  "eps_only", "eps_yoy_pct", "qrtrs_to_profitability",
  "gm_trend",
  // Valuation
  "price", "ps_now", "price_pct_of_52w_high",
  // Momentum
  "perf_52w_vs_spy", "composite_score",
  // Narrative
  "short_outlook", "key_risks", "full_outlook", "bull_eval", "bear_eval",
  "status",
  // Quality flags + audit
  "flags", "ai_analyzed_at",
] as const;

/**
 * Build the buy-time snapshot from the Level 0 fact store rather than the
 * legacy `companies` table:
 *   - identity / fundamentals / valuation → `api_universe_facts` (getEquityL0)
 *   - bull/bear + narrative → `ai_analysis`
 *
 * Maps each snapshot field name (kept in lock-step with theses.py) onto its
 * Level 0 source. Fields with no Level 0 equivalent (`composite_score`,
 * `rating`, `flags`, `gm_trend`, the revenue-consistency / opex / sm_rd
 * efficiency columns, `eps_yoy_pct`, `price_pct_of_52w_high`,
 * `perf_52w_vs_spy`, `qrtrs_to_profitability`, `r40_score`) are stored as
 * null in the frozen snapshot.
 */
async function buildSnapshot(ticker: string): Promise<Record<string, unknown>> {
  const supabase = getSupabase();
  const t = ticker.toUpperCase();

  const [equity, aiRes] = await Promise.all([
    getEquityL0(t),
    supabase
      .from("ai_analysis")
      .select(
        "bull_eval, bear_eval, short_outlook, key_risks, full_outlook, analyzed_at",
      )
      .eq("ticker", t)
      .maybeSingle(),
  ]);

  if (!equity) {
    // Caller should already have validated the ticker (buy() resolves the
    // price via getPrice first), so this is a hard error. Still throw a
    // recognisable shape so callers can decide whether to swallow.
    throw new Error(`buildSnapshot: no Level 0 facts for ${ticker}`);
  }

  const ai = (aiRes.data as Record<string, unknown> | null) ?? {};

  // Map each snapshot field to its Level 0 source value.
  const sourced: Record<string, unknown> = {
    // Identity / overview
    ticker: equity.ticker,
    company_name: equity.company_name,
    country: equity.country,
    sector: equity.sector,
    // Fundamentals
    rating: null,
    r40_score: null,
    rule_of_40: equity.rule_of_40,
    rev_growth_ttm_pct: equity.rev_growth_ttm_pct,
    rev_growth_qoq_pct: equity.rev_growth_qoq_pct,
    rev_cagr_pct: equity.rev_cagr_pct,
    rev_consistency_score: null,
    gross_margin_pct: equity.gross_margin_pct,
    operating_margin_pct: equity.operating_margin_pct,
    net_margin_pct: equity.net_margin_pct,
    net_margin_yoy_pct: null,
    fcf_margin_pct: equity.fcf_margin_pct,
    opex_pct_revenue: null,
    sm_rd_pct_revenue: null,
    eps_only: equity.eps_only,
    eps_yoy_pct: null,
    qrtrs_to_profitability: null,
    gm_trend: null,
    // Valuation
    price: equity.price,
    ps_now: equity.ps_now,
    price_pct_of_52w_high: null,
    // Momentum
    perf_52w_vs_spy: null,
    composite_score: null,
    // Narrative
    short_outlook: ai.short_outlook ?? null,
    key_risks: ai.key_risks ?? null,
    full_outlook: ai.full_outlook ?? null,
    bull_eval: ai.bull_eval ?? null,
    bear_eval: ai.bear_eval ?? null,
    status: equity.status,
    // Quality flags + audit
    flags: null,
    ai_analyzed_at: ai.analyzed_at ?? null,
  };

  const snapshot: Record<string, unknown> = {};
  for (const field of SNAPSHOT_FIELDS) {
    snapshot[field] = sourced[field] ?? null;
  }
  return snapshot;
}

/**
 * Record a thesis row for a freshly-executed BUY.
 *
 * Always captures the snapshot. If any of `thesis_text` /
 * `extend_signals` / `break_signals` is provided, `source='agent'`;
 * otherwise `source='auto'`. Marks any prior `active` row for the same
 * (agentId, ticker) as `superseded`.
 *
 * Returns the new thesis id, or null if the insert silently no-op'd
 * (e.g. RLS rejection — but the snapshot insert runs under the
 * service-role key, so this shouldn't happen in practice).
 */
export async function recordThesis(opts: {
  agentId: string;
  ticker: string;
  tradeId: number | null;
  portfolioId?: string | null;
  thesis?: ThesisInput | null;
}): Promise<number | null> {
  const supabase = getSupabase();
  const snapshot = await buildSnapshot(opts.ticker);

  // Supersede any prior active row for this (portfolio, ticker) — falls
  // back to (agent, ticker) when portfolioId isn't supplied. Both reach
  // the same rows during the 1:1 shim period.
  const supersedeBuilder = supabase
    .from("investment_theses")
    .update({
      status: "superseded",
      status_changed_at: new Date().toISOString(),
    })
    .eq("ticker", opts.ticker)
    .eq("status", "active");
  if (opts.portfolioId) {
    await supersedeBuilder.eq("portfolio_id", opts.portfolioId);
  } else {
    await supersedeBuilder.eq("agent_id", opts.agentId);
  }

  const text = opts.thesis?.thesis_text ?? null;
  const extend = opts.thesis?.extend_signals ?? null;
  const breakSig = opts.thesis?.break_signals ?? null;
  const source = text || extend || breakSig ? "agent" : "auto";

  const { data, error } = await supabase
    .from("investment_theses")
    .insert({
      agent_id: opts.agentId,
      portfolio_id: opts.portfolioId ?? opts.agentId,
      ticker: opts.ticker,
      trade_id: opts.tradeId,
      snapshot,
      thesis_text: text,
      extend_signals: extend,
      break_signals: breakSig,
      source,
      status: "active",
    })
    .select("id")
    .single();

  if (error || !data) {
    // Don't blow up the buy on a thesis-recording failure — log and move on.
    // Matches the Python side's exception-safe wrapper.
    console.error("recordThesis insert failed:", error);
    return null;
  }
  return (data as { id: number }).id;
}

/**
 * Flip all non-closed theses for (agentId, ticker) to `status='closed'`.
 * Called from `sell()` after the position is zeroed out. Idempotent —
 * a no-op if there are no matching rows (e.g. legacy positions opened
 * before migration 020).
 */
export async function closeThesesForPosition(opts: {
  agentId: string;
  ticker: string;
  portfolioId?: string | null;
}): Promise<void> {
  const supabase = getSupabase();
  const now = new Date().toISOString();
  const builder = supabase
    .from("investment_theses")
    .update({
      status: "closed",
      status_changed_at: now,
      closed_at: now,
    })
    .eq("ticker", opts.ticker)
    .neq("status", "closed");
  const filtered = opts.portfolioId
    ? builder.eq("portfolio_id", opts.portfolioId)
    : builder.eq("agent_id", opts.agentId);
  const { error } = await filtered;
  if (error) {
    console.error("closeThesesForPosition update failed:", error);
  }
}
